File: pipeline/use-predictions/save-predictions.R
save_predictions <- function(predictions, prod_run = FALSE){

  if (prod_run == TRUE){

    googledrive::drive_auth(
      path = paste0(here(), "/footy-tipper-c5bcb9639ee2.json"),
      scopes = "https://www.googleapis.com/auth/drive"
    )
    write.csv(x = predictions, file = "predictions.csv")

    # Upload the file
    drive_upload(media = "predictions.csv",
                 path = "footy-tipping-predictions/",
                #  name = paste0("round", unique(predictions$round_id), "_",
                #                unique(predictions$competition_year), ".csv"),
                name = "test",
                 type = NULL,
                 overwrite = TRUE)

    unlink("predictions.csv")

  }

}

get_punt_thresholds <- function(predictions){

  punt_thresholds <- predictions %>%
    mutate(home_odds_thresh = 1/home_team_win_prob,
           away_odds_thresh = 1/home_team_lose_prob) %>%
    select(team_home, home_odds_thresh, team_away, away_odds_thresh)
  
  return(punt_thresholds)
  
}

levons_picks <- function(predictions, prod_run = FALSE){
  
  punt_thresholds <- predictions %>%
    mutate(home_odds_thresh = 1/home_team_win_prob,
           away_odds_thresh = 1/home_team_lose_prob)
  
  home_picks <- punt_thresholds %>%
    filter(home_team_result == 'Win') %>%
    select(team_home, team_head_to_head_odds_home, home_odds_thresh) %>%
    rename(team = team_home, price = team_head_to_head_odds_home, price_min = home_odds_thresh)
  
  away_picks <- punt_thresholds %>%
    filter(home_team_result == 'Loss') %>%
    select(team_away, team_head_to_head_odds_away, away_odds_thresh) %>%
    rename(team = team_away, price = team_head_to_head_odds_away, price_min = away_odds_thresh)
  
  levons_picks <- bind_rows(home_picks, away_picks) %>%
    filter(price > (price_min*1.15)) %>%
    mutate_if(is.numeric, dollar)
  
  if (prod_run == TRUE){
  
    write.csv(x = levons_picks, file = "levons_picks.csv")
    
    # Upload the file
    drive_upload(media = "levons_picks.csv",
                 path = "footy-tipping-predictions/levons_picks",
                 name = paste0("levons_picks_round", unique(predictions$round_id), "_", unique(predictions$competition_year), ".csv"), 
                 type = NULL,
                 overwrite = TRUE)
    
    unlink("levons_picks.csv")
  
  }
  
  return(levons_picks)
  
}

File: pipeline/use-predictions/save_predictions.py
import os
import pandas as pd
from google.oauth2.credentials import Credentials
from google_auth_oauthlib.flow import InstalledAppFlow
from google.auth.transport.requests import Request
from googleapiclient.discovery import build
from googleapiclient.http import MediaFileUpload

def save_predictions(predictions, prod_run=False):
    if prod_run:
        # OAuth 2.0 setup
        creds = None
        scopes = ['https://www.googleapis.com/auth/drive']

        if creds and creds.expired and creds.refresh_token:
            creds.refresh(Request())
        else:
            flow = InstalledAppFlow.from_client_secrets_file('path_to_your_credentials.json', scopes)
            creds = flow.run_local_server(port=0)
        
        # Build the Google Drive service
        drive_service = build('drive', 'v3', credentials=creds)

        predictions.to_csv('predictions.csv', index=False)
        
        # Upload the file
        file_metadata = {
            'name': f"round{predictions['round_id'].unique()[0]}_{predictions['competition_year'].unique()[0]}.csv",
            'parents': ['your_folder_id']
        }
        media = MediaFileUpload('predictions.csv', mimetype='text/csv')
        drive_service.files().create(body=file_metadata, media_body=media, fields='id').execute()


def tipper_picks(predictions, prod_run=False):
    predictions['home_odds_thresh'] = 1 / predictions['home_team_win_prob']
    predictions['away_odds_thresh'] = 1 / predictions['home_team_lose_prob']
    
    home_picks = predictions[predictions['home_team_result'] == 'Win'][['team_home', 'team_head_to_head_odds_home', 'home_odds_thresh']].copy()
    home_picks.rename(columns={'team_home': 'team', 'team_head_to_head_odds_home': 'price', 'home_odds_thresh': 'price_min'}, inplace=True)
    
    away_picks = predictions[predictions['home_team_result'] == 'Loss'][['team_away', 'team_head_to_head_odds_away', 'away_odds_thresh']].copy()
    away_picks.rename(columns={'team_away': 'team', 'team_head_to_head_odds_away': 'price', 'away_odds_thresh': 'price_min'}, inplace=True)
    
    tipper_picks = pd.concat([home_picks, away_picks])
    tipper_picks = tipper_picks[tipper_picks['price'] > (tipper_picks['price_min'] * 1.15)]
    
    if prod_run:
        # OAuth 2.0 setup
        creds = None
        scopes = ['https://www.googleapis.com/auth/drive']

        if creds and creds.expired and creds.refresh_token:
            creds.refresh(Request())
        else:
            flow = InstalledAppFlow.from_client_secrets_file('path_to_your_credentials.json', scopes)
            creds = flow.run_local_server(port=0)
        
        # Build the Google Drive service
        drive_service = build('drive', 'v3', credentials=creds)

        tipper_picks.to_csv('tipper_picks.csv', index=False)
        
        # Upload the file
        file_metadata = {
            'name': f"tipper_picks_round{predictions['round_id'].unique()[0]}_{predictions['competition_year'].unique()[0]}.csv",
            'parents': ['your_folder_id']
        }
        media = MediaFileUpload('tipper_picks.csv', mimetype='text/csv')
        drive_service.files().create(body=file_metadata, media_body=media, fields='id').execute()

        return tipper_picks
